### PR TITLE
Parallelized BDTs in TMVA

### PR DESCRIPTION
**Pull request by acarnes**
Parallelized Boosted Decision Trees for regression in TMVA.

### CODE REVIEW

File: tmva/tmva/src/DecisionTree.cxx [hunk 1]
```diff
@@ -319,15 +412,280 @@ UInt_t TMVA::DecisionTree::BuildTree( const std::vector<const TMVA::Event*> & ev
...
    else Log() << kFATAL << ":<BuildTree> eventsample Size == 0 " << Endl;
 
+
+   // sum up the totals
+   // sig and bkg for classification
+   // err and err2 for regression
+
+   // #### Set up prerequisite info for multithreading
+   UInt_t nPartitions = fNumCPUs;
+   auto seeds = ROOT::TSeqU(nPartitions);
```
Comment (xvallspl) [inline tmva/tmva/src/DecisionTree.cxx]: TThreadExecutor has mechanisms to implement chunking, you may want to check them. See it in action here: https://github.com/root-project/root/blob/master/math/mathcore/src/FitUtil.cxx#L370 where we have a function called setAutomaticChunking that makes a guess (based on observations, but almost arbitrary)of the number of chunks to take. It's better than chunking by number of logical processors in most of the cases.

File: tmva/tmva/src/DecisionTree.cxx [hunk 2]
```diff
@@ -68,6 +68,7 @@ more signal respective background events from the training sample.
 #include "TRandom3.h"
 #include "TMath.h"
 #include "TMatrix.h"
+#include "TStopwatch.h"
```
Comment (xvallspl) [inline tmva/tmva/src/DecisionTree.cxx]: Is this necessary? You aren't using any of the TStopwatch functions.

File: tmva/tmva/src/LossFunction.cxx [hunk 3]
```diff
@@ -41,7 +41,8 @@ Huber Loss Function.
 
 #include "Rtypes.h"
 #include "TMath.h"
-
+#include "TStopwatch.h"
```
Comment (xvallspl) [inline tmva/tmva/src/LossFunction.cxx]: Is this necessary? You're not using any of the TStopwatch methods. This issue happens in several files

File: tmva/tmva/src/LossFunction.cxx [hunk 4]
```diff
@@ -79,8 +80,36 @@ void TMVA::HuberLossFunction::Init(std::vector<LossFunctionEventInfo>& evs){
...
 
+      for(Int_t i=start; i<end; i++)
+      {
+          sumOfWeightsN+=evs[i].weight;
+      }
+
+      return sumOfWeightsN;
+   };
+
+   auto sumOfWeightsN = fPool->MapReduce(f, seeds, redfunc, nPartitions);
```
Comment (xvallspl) [inline tmva/tmva/src/LossFunction.cxx]: Probably shouldn't have the same name than the return type from the lambda

**Comment by xvallspl** [comment]
Can you make clang-format happy by fixing the formatting? :)

File: tmva/tmva/inc/TMVA/DecisionTree.h [hunk 5]
```diff
@@ -187,9 +192,35 @@ namespace TMVA {
...
+         gSystem->GetSysInfo(&s);
+         UInt_t ncpu  = s.fCpus;
+         return ncpu;
+      };
+
+      // #### root multithreading object
+      // #### had to define as pointer so that we can initialize the number of CPUs to use AFTER
+      // #### getting information from the user. Needed to do this to automate timing vs nCPUs during
+      // #### development.
+      std::unique_ptr<ROOT::TThreadExecutor> fPool = std::unique_ptr<ROOT::TThreadExecutor>(nullptr);
```
Comment (acarnes) [inline tmva/tmva/inc/TMVA/DecisionTree.h]: I mainly created options for nthreads so that I could do the bench marking studies. The option is defaulted to use all of the available CPUs. If you would like the user to be able to set the number of CPUs then let them know about the option and their choice will override the default. If not, just don't let them know about it and it will be set automatically. The way this is coded is flexible, though I'm not sure how it differs from the default constructor.

Anyways, I would argue to default the threading option to the desired behavior for an uninformed user, but allow an informed user to have greater control.

File: tmva/tmva/src/LossFunction.cxx (hunk 3 above)
Comment (acarnes) [inline tmva/tmva/src/LossFunction.cxx]: I accidentally left these includes in. It's a relic from bench marking the timing vs nthreads.

File: tmva/tmva/src/LossFunction.cxx (hunk 4 above)
Comment (acarnes) [inline tmva/tmva/src/LossFunction.cxx]: Whoops. Not the best naming, but shouldn't affect anything.

**Comment by acarnes** [comment]
How do I see the changes I need to make to make clang happy?

**Comment by xvallspl** [comment]
Nevermind, It turns out that TMVA follows different code conventions. I don't know if it uses Doxygen documentation though.

File: tmva/tmva/src/LossFunction.cxx (hunk 4 above)
Comment (xvallspl) [inline tmva/tmva/src/LossFunction.cxx]: Yeah, just suggesting.

**Comment by ashlaban** [comment]
@phsft-bot build!

File: tmva/tmva/src/MethodBDT.cxx [hunk 6]
```diff
@@ -408,6 +419,15 @@ void TMVA::MethodBDT::DeclareOptions()
       fSepTypeS = "GiniIndex";
    }
 
+   // #### Number of CPUs to use in fPool multithreading, if you want the user to be able to set this
+   // rather than use all of the available CPUs, let them know about this option in the User's Guide or tutorials
+   // I just used it to time the algorithm vs number of CPUs.
+   #ifndef R__USE_IMT
+   DeclareOptionRef(fNumCPUs = 1, "NumCPUs", "Number of CPUs to use in multithreading.");
```
Comment (ashlaban) [inline tmva/tmva/src/MethodBDT.cxx]: IIUC this should probably be renamed to fNumThreads and "NumThreads" to better reflect that we are using multithreading.

File: tmva/tmva/src/MethodBDT.cxx [hunk 7]
```diff
@@ -2463,6 +2536,9 @@ const std::vector<Float_t> & TMVA::MethodBDT::GetRegressionValues()
       evT->SetTarget(0, rVal/Double_t(count) );
    }
    else if(fBoostType=="Grad"){
+      // #### Can parallelize this loop over the trees...
+      // #### but this doesn't take much time now that the evaluation bug
+      // #### has been fixed
```
Comment (ashlaban) [inline tmva/tmva/src/MethodBDT.cxx]: Remove comment

File: tmva/tmva/src/MethodBase.cxx [hunk 8]
```diff
@@ -762,11 +762,22 @@ void TMVA::MethodBase::AddRegressionOutput(Types::ETreeType type)
...
+
    for (Int_t ievt=0; ievt<nEvents; ievt++) {
+
       Data()->SetCurrentEvent(ievt);
       std::vector< Float_t > vals = GetRegressionValues();
       regRes->SetValue( vals, ievt );
-      timer.DrawProgressBar( ievt );
+
+      // Only draw the progress bar once in a while, doing this every event causes the evaluation to be ridiculously slow
+      if(ievt % drawProgressEvery == 0 || ievt==nEvents-1) timer.DrawProgressBar( ievt );
```
Comment (ashlaban) [inline tmva/tmva/src/MethodBase.cxx]: Can we extract this into a separate PR? It's an independent bugfix that is _very_ nice regardless of what eventually happens with this PR.

File: tmva/tmva/src/MethodBDT.cxx (hunk 6 above)
Comment (dpiparo) [reply tmva/tmva/src/MethodBDT.cxx]: Shouldn't this be working on parallel implicitly w/o the need of specifying explicitly the number of cpus?

Comment (ashlaban) [reply tmva/tmva/src/MethodBDT.cxx]: After further contemplation: Yes, it most probably should.

Can we use imt for our benchmarking as well? I.e. Can one specify the size of the thread pool?

Comment (acarnes) [reply tmva/tmva/src/MethodBDT.cxx]: @dpiparo I mainly created options for nthreads so that I could do the bench marking studies. However, the option is defaulted to use all of the available CPUs. If we would like the user to be able to set the number of CPUs then let them know about the option and their choice will override the default. If not, just don't let them know about it and it will be set automatically. The way this is coded is flexible.

Anyways, I would argue to default the threading option to the desired behavior for an uninformed user, but allow an informed user to have greater control -- the current set up.

Comment (acarnes) [reply tmva/tmva/src/MethodBDT.cxx]: @ashlaban fNumThreads and numThreads are indeed better names. In regard to your other comment, there are options to specify the size of the threadpool if I remember correctly, but I just had it set up to use the number of threads as the pool size. Maybe @xvallspl can comment.

Comment (xvallspl) [reply tmva/tmva/src/MethodBDT.cxx]: @acarnes Yes, suggested in here https://github.com/root-project/root/pull/723#discussion_r125260915 (there's also a couple more comments I would like to be addressed before merging)

Your user can set the number of threads in their application with ROOT::EnableImplicitMT(), and then your class would ideally rely on that value, by default as much as your computer has available.

**Comment by xvallspl** [comment]
@acarnes ping? What did you agree with Sergei? Is someone working on this?

**Comment by acarnes** [comment]
@xvallspl My analysis has slowed down. I will take care of the needed updates early next week.

**Comment by lmoneta** [comment]
Hi Andrew, 

 I see in your code you create inside your TThreadExecutor instance. I think it is better to uleave this to TMVA::Config. We have changed this for the DNN, and in the latest master we have now a global TThreadExecutor created in the singleton TMVA::Config class and can be access via TMVA::gConfig()::GetThreadExecutor(). 

See for example https://github.com/root-project/root/blob/master/tmva/tmva/inc/TMVA/DNN/Architectures/Cpu/CpuMatrix.h#L105

If you need to have more functions (e.g. number of threads or cpu) we can add them in TMVA::Config. 
I am planning to add a function TMVA::Config::GetNCpu() , it is already in our dev branch for the DNN. 

Lorenzo

**Comment by lmoneta** [comment]
Hi Andrew

Can you please fix the conflicts so we can proceed with the testing and then merging of this PR.
I would suggest to change the following..
Instead of having the TThreadExecutor object as data member of MethodBDT and LossFunction, you should use it from TMVA::Config. 
So basically replace the fPool in MethodBDT.cxx and LossFunction.cxx with 
TMVA::Config::Instance().GetThreadExecutor()

Thank you for applying these changes. If you don't have time to do them, please let us know and we will eventually do it

 Best Regards

 Lorenzo

**Comment by acarnes** [comment]
Hi Xavi,

I will start to fix the conflicts tomorrow.

Best,
    Andrew

**Comment by lmoneta** [comment]
Hi Andrew, 

Thank you for the changes. These are perfect !
The only thing is that this PR has still two conflicts with the master, in LoassFunction and MethodBDT. 
It would be great if you can fix them so we can test it and then merge

Cheers

 Lorenzo

**Comment by ashlaban** [comment]
Should we merge this after the next dev release?

**Comment by lmoneta** [comment]
@phsft-bot build

**Comment by lmoneta** [comment]
This PR is now replaced by PR #1781
